### PR TITLE
Add "priority" column to ACLs and support ACL rule precedence 

### PR DESCRIPTION
**Pull request by seamuslee001**
Overview
----------------------------------------
This extends the work done by #26041 by adding in the concept of a *priority* to ACL rules. This allows for allow or deny rules to take precedence by the priority they are given.

Before
----------------------------------------
Deny rules always override allow rules without any real flexibility

After
----------------------------------------
Flexibility

ping @colemanw

### CODE REVIEW

File: CRM/ACL/BAO/ACL.php [hunk 1]
```diff
@@ -530,22 +530,37 @@ private static function loadDenyIDs(int $contactID, string $tableName, int $type
...
-SELECT   a.operation, a.object_id
+SELECT   a.operation, a.object_id, a.weight
   FROM   civicrm_acl_cache c, civicrm_acl a
  WHERE   c.acl_id       =  a.id
    AND   a.is_active    =  1
    AND   a.object_table = %1
    AND   a.id        IN ( $aclKeys )
    AND   a.deny         = 1
-GROUP BY a.operation,a.object_id
+GROUP BY a.operation,a.object_id, a.weight
```
Comment (colemanw) [inline CRM/ACL/BAO/ACL.php:540]: Why are you grouping by `weight`? I would think you would want to order by weight instead.

File: CRM/ACL/BAO/ACL.php [hunk 2]
```diff
@@ -530,22 +530,37 @@ private static function loadDenyIDs(int $contactID, string $tableName, int $type
     $aclKeys = array_keys($acls);
     $aclKeys = implode(',', $aclKeys);
     $query = "
-SELECT   a.operation, a.object_id
+SELECT   a.operation, a.object_id, a.weight
```
Comment (colemanw) [inline CRM/ACL/BAO/ACL.php:533]: I'm a little concerned about this query causing a crash in an upgrade situation when the column hasn't been added yet. `DAO::getSupportedFields()` is your friend.

File: templates/CRM/ACL/Page/ACL.tpl [hunk 3]
```diff
@@ -31,6 +31,7 @@
               <th>{ts}Description{/ts}</th>
               <th>{ts}Enabled?{/ts}</th>
               <th>{ts}Mode{/ts}</th>
+              <th>{ts}Weight{/ts}</th>
```
Comment (colemanw) [inline templates/CRM/ACL/Page/ACL.tpl:34]:
```suggestion
              <th>{ts}Priority{/ts}</th>
```

File: CRM/Upgrade/Incremental/php/FiveSixtyFour.php [hunk 4]
```diff
@@ -28,6 +28,7 @@ class CRM_Upgrade_Incremental_php_FiveSixtyFour extends CRM_Upgrade_Incremental_
    *   The version number matching this function name
    */
   public function upgrade_5_64_alpha1($rev): void {
+    $this->addTask('Add weight column onto ACL table', 'addColumn', 'civicrm_acl', 'priority', 'int NOT NULL DEFAULT 0');
```
Comment (colemanw) [inline CRM/Upgrade/Incremental/php/FiveSixtyFour.php:31]:
```suggestion
    $this->addTask('Add priority column onto ACL table', 'addColumn', 'civicrm_acl', 'priority', 'int NOT NULL DEFAULT 0');
```

File: CRM/ACL/BAO/ACL.php [hunk 5]
```diff
@@ -481,31 +479,47 @@ protected static function loadPermittedIDs(int $contactID, string $tableName, in
     $acls = CRM_ACL_BAO_Cache::build($contactID);
     $aclKeys = array_keys($acls);
     $aclKeys = implode(',', $aclKeys);
+    $select = 'a.operation, a.object_id,a.deny';
+    $orderBy = 'a.object_id';
+    $hasPriorty = FALSE;
+    if (array_key_exists('priority', CRM_ACL_BAO_ACL::getSupportedFields())) {
+      $select .= ',a.priority';
```
Comment (colemanw) [inline CRM/ACL/BAO/ACL.php:486]: I don't think this needs to be selected.

File: CRM/ACL/BAO/ACL.php [hunk 6]
```diff
@@ -481,31 +479,47 @@ protected static function loadPermittedIDs(int $contactID, string $tableName, in
...
+      $orderBy .= ',a.priority';
+      $hasPriority = TRUE;
```
Comment (colemanw) [inline CRM/ACL/BAO/ACL.php:488]: Unused variable?

**Comment by colemanw** [comment]
Ok great, this has passed review and tests are passing.

**Comment by TobiasVoigt** [comment]
Hi there,

today I tested the newest implementation of the weighting functionality in CiviCRM 5.65.0 and found that I wasn't able to set up my initial usecase due to what seems to be a bug.

My usecase included three ACLs:

ACL rule 1: Allow access to all custom field groups for all authenticated users → Priority 1
ACL rule 2: Disallow access to the privileged set of custom fields for all authenticated users → Priority 2
ACL rule 3: Allow access to the privileged set of custom fields for a privileged role → Priority 3

I used the new weighting functionality but didn't get the results I expected. I expected that priority 3 would have precedence over 2 (and 2 over 1), in other words: The higher the number, the higher the priority. But with this setup all users where able to access all custom fields.

I then made some tests and realized:

1. When I create a test setup with several rules that all refer to ONE SPECIFIC custom field group the weighting logic works  like I expected: The higher number has the priority.
2. When I create a test setup with several rules that all refer to ALL custom field groups, the weighting logic seems to be flipped: The smaller number has the higher priority.

So my guess is that somewhere in the ACL code the weighting logic is somehow flipped for "All custom field groups". This would explain why mixing rules for "All custom field groups" and rules for specific custom field groups in my test setup didn't work.

**To my regret, this means I won't be able to show the new functionality to a live audience at CiviCamp in Leipzig next monday (2023-09-11). If there's any chance someone would be able to deliver a quick fix that would be absolutely awesome.** 

In any case - I hope my analysis makes sense and shed some light onto a potential bug. Let me know! :)

Best regards,

Tobias

**Pull request merged**